### ISSUE TITLE
Initialize media list for lights and camera sampling 

### ISSUE THREAD

**Issue by Biart95** [appleseedhq/appleseed#1485]
Currently, we can properly render volumes only if camera is located outside of any volume. In order to place camera inside a volume, the media list have to be properly initialized with all the surrounding media before path tracing starts.

The same stands for lights and light tracing, as well as for photon tracing.

**Comment by Biart95** [comment]
Just initializing media list, though, is not sufficient. We also need to compute transmission and in-scattering radiance for the initial ray, which now can be done only during the path tracing itself.

**Comment by dictoon** [comment]
I'm a bit surprised, typically knowing which volume(s) a ray starts is usually enough...

**Comment by Biart95** [comment]
Our raymarching code is placed into `PathTracer`. This means that the initial ray cannot be raymarched, since it is not processed in a `PathTracer` but directly in sample renderer.

The design of our renderer is clear: primary ray either hits any surface or the environment, and then the corresponding shader is evaluated. However, in case of volume rendering there is a problem here. 

Let's say we implemented multiple scattering. And we got a scene where the camera is located inside a volume. Logically, the primary ray can be outscattered even before it is able to reach any surface, i.e. even before the surface shader is evaluated. But that is contradictary! Without evaluating the shader we don't even know if multiple scattering is enabled, not to mention that the parameters of raymarcher are missing.

For single scattering and homogeneous volumes we can integrate the lighting postfactum, in the very beginning of `PathTracer::trace`, but it is obviously a cheating.

**Comment by dictoon** [comment]
You're right. We probably need to alter the design slightly such that the primary ray is being passed to the path tracer, and is then traced like any other ray of the path. Would that be enough to solve the issue cleanly?

**Comment by Biart95** [comment]
Yes, if this change will be possible than it is absolutely feasible task!

**Comment by dictoon** [comment]
I think it's indeed something we can change.
